File: medium/133-Clone-Graph/DFS.kt
/**
 * Definition for a Node.
 * class Node(var `val`: Int) {
 *     var neighbors: ArrayList<Node?> = ArrayList<Node?>()
 * }
 */

class Solution {
    fun cloneGraph(node: Node?): Node? {
        if (node == null) return null
        
        var stack = ArrayDeque<Node>()
        stack.push(node)
        
        var cloned = HashMap<Int, Node>()
        cloned[node.`val`] = Node(node.`val`)
        
        while (!stack.isEmpty()) {
            val crr = stack.pop()
            for (adj in crr.neighbors) {
                if (adj == null) continue
                if (!(adj.`val` in cloned)) {
                    cloned[adj.`val`] = Node(adj.`val`)
                    stack.push(adj)
                }
                cloned[crr.`val`]?.neighbors?.add(cloned[adj.`val`])
            }
        }
        
        return cloned[node.`val`]
    }
}

File: medium/133-Clone-Graph/DFS.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""


class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node is None:
            return None

        cloned = {}
        stack = [node]
        cloned[node.val] = Node(node.val)

        while stack:
            orig = stack.pop()
            for adj in orig.neighbors:
                if adj.val not in cloned:
                    stack.append(adj)
                    cloned[adj.val] = Node(adj.val)
                cloned[orig.val].neighbors.append(cloned[adj.val])

        return cloned[node.val]
